File: LiteDb.OneBella/UseCases/RunSql.fs
module OneBella.UseCases.RunSql

open System
open System.Diagnostics
open System.IO
open System.Text
open System.Text.RegularExpressions
open System.Threading
open LiteDB
open LiteDb.Studio.Avalonia.Core
open OneBella.Core
open OneBella.Core.Rop

type T =
    { Stopwatch: Stopwatch
      Query: string
      Db: unit -> LiteDatabase
      Token: CancellationToken }

let create sw query db token =
    { Stopwatch = sw
      Query = query
      Db = db
      Token = token }

let removeComments sql =
    use reader = new StringReader(sql)
    let sb = StringBuilder()

    let rec read () =
        match reader.ReadLine() with
        | null -> sb
        | s when s.TrimStart().StartsWith("--") -> read ()
        | a ->
            sb.AppendLine(a) |> ignore
            read ()

    read().ToString()

let findTableName sql =
    let getPattern (actualSql: string) =
        if actualSql.Contains("INSERT") then
            Some @"INSERT\s+INTO\s+(?<table>\w+).*"
        elif actualSql.Contains("UPDATE") then
            Some @"UPDATE\s+(?<table>\w+).*"
        elif actualSql.Contains("DELETE") then
            Some @"DELETE\s+(?<table>\w+).*"
        elif actualSql.Contains("SELECT") then
            Some @"SELECT.*FROM\s+(?<table>\w+).*"
        else
            None


    sql
    |> removeComments
    |> fun s -> s.ToUpperInvariant().Trim()
    |> fun sql -> sql, getPattern sql
    |> fun (sql, o) ->
        match o with
        | None -> None
        | Some r ->
            let regex = Regex.Match(sql, r)
            if regex.Success then
                let g = regex.Groups.["table"]
                Some g.Value
            else
                None


let run (req: T) =
    let db = req.Db()
    req.Stopwatch.Restart()
    use reader =  DbUtils.Exec(db, req.Query |> removeComments);
    let bsonValues = DbUtils.ReadResult(reader, req.Token) |> Seq.map BVal.Create |> Seq.toArray
    //|> Seq.toArray
    req.Stopwatch.Stop()
    bsonValues

File: LiteDb.Studio.Avalonia/UseCases/RunSql.cs
using System.Diagnostics;
using System.Text;
using System.Text.RegularExpressions;
using LiteDB;
using LiteDb.Studio.Avalonia.Core;

namespace LiteDb.Studio.Avalonia.UseCases;

public static class RunSql
{
    public record T
    {
        public Stopwatch Stopwatch { get; init; } = new Stopwatch();
        public string Query { get; init; } = "";
        public Func<LiteDatabase> Db { get; init; } = default!;
        public CancellationToken Token { get; init; }
    }

    public static T Create(Stopwatch sw, string query, Func<LiteDatabase> db, CancellationToken token) =>
        new T { Stopwatch = sw, Query = query, Db = db, Token = token };

    public static string RemoveComments(string sql)
    {
        using var reader = new StringReader(sql);
        var sb = new StringBuilder();
        string? line;
        while ((line = reader.ReadLine()) != null)
        {
            if (!line.TrimStart().StartsWith("--"))
                sb.AppendLine(line);
        }
        return sb.ToString();
    }

    public static string? FindTableName(string sql)
    {
        string? GetPattern(string actualSql)
        {
            if (actualSql.Contains("INSERT")) return @"INSERT\s+INTO\s+(?<table>\w+).*";
            if (actualSql.Contains("UPDATE")) return @"UPDATE\s+(?<table>\w+).*";
            if (actualSql.Contains("DELETE")) return @"DELETE\s+(?<table>\w+).*";
            if (actualSql.Contains("SELECT")) return @"SELECT.*FROM\s+(?<table>\w+).*";
            return null;
        }

        var cleaned = RemoveComments(sql).ToUpperInvariant().Trim();
        var pattern = GetPattern(cleaned);
        if (pattern == null) return null;

        var match = Regex.Match(cleaned, pattern);
        return match.Success ? match.Groups["table"].Value : null;
    }

    public static BValType[] Run(T req)
    {
        var db = req.Db();
        req.Stopwatch.Restart();
        using var reader = DbUtils.Exec(db, RemoveComments(req.Query));
        var bsonValues = DbUtils.ReadResult(reader, req.Token).Select(BVal.Create).ToArray();
        req.Stopwatch.Stop();
        return bsonValues;
    }
}